### PR TITLE
parser: verify and fix offAfterPRI calculation

### LINKED ISSUE #6017: [General]: Parser: verify and, if needed, fix `offAfterPRI` offset calculation (internal field)
### 📝 Description / What's on your mind:

`offAfterPRI` is an internal `msg_t` field used by parser modules to know the byte offset immediately after the initial PRI. The open question is: can this field be miscalculated? If yes, implement a fix.

Context from an older tracker note:

* It is set in `parser.c` and used by some parser modules.
* Suspicion that the offset may be calculated incorrectly in corner cases.
* Quote (2015-05-27) from https://github.com/rsyslog/rsyslog/issues/381: "yes, it's being used... The 'general parser' sets it, because it processes PRI. The parser modules use it to know where to begin to look at. What remains is the question if it is sometimes miscalculated."

Define expected semantics to check against:

* If a message starts with a valid PRI, `offAfterPRI` must equal the index of the first byte after the closing `>` of that PRI.
* Invariants when set:

  * `0 <= offAfterPRI <= msg_len`
  * `raw[0] == '<'`
  * `raw[offAfterPRI - 1] == '>'`
  * The digits between `<` and `>` are 1..3 ASCII digits. Range validation of PRI value is optional for offset calc, but must not cause OOB access.

Proposed work items:

1. Audit where `offAfterPRI` is assigned and where it is consumed. Ensure it is computed exactly once in the "general parser" and only read elsewhere.
2. Extract a small helper (internal): `compute_off_after_pri(const unsigned char *buf, size_t len)` returning `int` offset or `-1`. Use strict bounds checks and no undefined behavior on malformed inputs.
3. Add assertions and defensive checks when setting the field:

   * Reject offsets outside `[0..len]`.
   * Only set non-negative on syntactically valid `<digits>` terminated by `>`.
4. Tests (unit/integration) for at least:

   * RFC3164 example with PRI.
   * RFC5424 example with PRI.
   * Missing `>`.
   * Non-digit between `<` and `>`.
   * More than 3 digits.
   * No leading `<`.
   * Leading NULs or binary octets before `<`.
   * Very short buffers (len 0..4).
5. If miscalculation is found, fix logic and keep behavior consistent across all parser modules.
6. Optional: tighten the inline Doxygen on the field to state the exact semantics and invariants above.

Acceptance criteria:

* Clear decision recorded: field is correctly computed in all tested cases, or a fix is merged.
* Centralized computation via the helper; no duplicate or module-local re-parsing of PRI.
* Tests prove correctness for valid and malformed inputs.
* Field remains internal; no public docs changes. Only inline Doxygen may be updated.

### PR DESCRIPTION
**Pull request by rgerhards**
closes https://github.com/rsyslog/rsyslog/issues/6017

### Why:
The internal `offAfterPRI` field tracks the offset in raw messages immediately after the PRI. This was inconsistently calculated across modules (e.g. `imuxsock` omitted the closing `>`) and was prone to parsing invalid strings (e.g. `<>`) as valid PRI offsets. This caused misalignments and potential out-of-bounds risks in downstream parser modules.

### Impact:
Stabilizes syslog parsing; downstream modules consistently receive accurate raw message text.

### Before/After:
`offAfterPRI` was inconsistently calculated or misaligned on malformed/special inputs; now it is centrally validated and correct.

### Technical Overview:
- Extracted the PRI offset logic into a strict static helper `compute_off_after_pri` in `runtime/parser.c` to parse 1..3 digits between `<` and `>`.
- Refactored `ParsePRI` to use this helper.
- Enhanced `MsgSetAfterPRIOffs` in `runtime/msg.c` with defensive assertions to validate offsets and enclosing brackets.
- Updated the legacy `imuxsock` parser in `plugins/imuxsock/imuxsock.c` to set the correct `offs + 1` offset when the closing `>` is present.
- Created a pure C unit test `tests/unit/parser_pri_test.c` checking 10 distinct RFC3164/RFC5424 corner cases.

With the help of AI-Agents: Antigravity

### CODE REVIEW

**Comment by rgerhards** [comment]
@codex review

File: plugins/imfifo/imfifo.c [hunk 1]
```diff
@@ -0,0 +1,448 @@
...
+        ABORT_FINALIZE(RS_RET_IO_ERROR);
+    } else if (nRead == 0) {
+        /* EOF reached (should not happen if opened O_RDWR) */
+        ABORT_FINALIZE(RS_RET_EOF);
+    }
+
+    for (ssize_t i = 0; i < nRead; ++i) {
+        if (buf[i] == '\n') {
+            CHKiRet(enqLine(pInst));
+            CHKiRet(rsCStrTruncate(pInst->ppCStr, cstrLen(pInst->ppCStr))); /* Reset buffer */
```
[the remark this answers is not in the document]
Comment (rgerhards) [reply plugins/imfifo/imfifo.c:274]: Not changed: rsCStrTruncate removes the requested number of bytes from the end; passing cstrLen() clears the accumulator. This matches the existing string API behavior and the imfifo tests pass.

File: plugins/imuxsock/imuxsock.c [hunk 2]
```diff
@@ -1107,7 +1107,7 @@ static rsRetVal SubmitMsg(uchar *pRcv, int lenRcv, lstn_t *pLstn, struct ucred *
         parser.SanitizeMsg(pMsg);
         lenMsg = pMsg->iLenRawMsg - offs; /* SanitizeMsg() may have changed the size */
         msgSetPRI(pMsg, pri);
-        MsgSetAfterPRIOffs(pMsg, offs);
+        MsgSetAfterPRIOffs(pMsg, (offs < pMsg->iLenRawMsg && pMsg->pszRawMsg[offs] == '>') ? offs + 1 : 0);
```
[the remark this answers is not in the document]
Comment (rgerhards) [reply plugins/imuxsock/imuxsock.c]: Done. imuxsock now requires a PRI with 1..3 digits, a closing >, and pri <= LOG_MAXPRI before setting afterPRI; malformed values keep LOG_PRI_INVLD and afterPRI 0.

File: runtime/parser.c [hunk 3]
```diff
@@ -514,38 +514,71 @@ static rsRetVal SanitizeMsg(smsg_t *pMsg) {
...
-    int lenMsg;
     DEFiRet;
 
     /* pull PRI */
-    lenMsg = pMsg->iLenRawMsg;
-    msg = pMsg->pszRawMsg;
-    pri = DEFUPRI;
     if (pMsg->msgFlags & NO_PRI_IN_RAW) {
         /* In this case, simply do so as if the pri would be right at top */
+        msgSetPRI(pMsg, DEFUPRI);
```
[the remark this answers is not in the document]
Comment (rgerhards) [reply runtime/parser.c]: Done. Removed the DEFUPRI overwrite so NO_PRI_IN_RAW preserves the input-provided PRI and only sets afterPRI to 0.

File: plugins/imfifo/imfifo.c [hunk 4]
```diff
@@ -0,0 +1,448 @@
...
+        } else {
+            CHKiRet(cstrAppendChar(pInst->ppCStr, buf[i]));
```
[the remark this answers is not in the document]
Comment (rgerhards) [reply plugins/imfifo/imfifo.c]: Done via the updated stack. imfifo caps accumulation at maxMessageSize, logs one truncation notice per overlong line, and the imfifo test covers the truncation oracle.

**Comment by rgerhards** [comment]
Rebased #7037 onto current upstream/main (26627f5a3) and removed the stale imfifo stack from #7029. Remaining diff is only the PRI/offAfterPRI parser work: imuxsock, runtime msg/parser, tests/Makefile.am, and tests/unit/parser_pri_test.c.

Local validation on new tip 10fe8ba58:
- local Cubic review: clean after fixing its NO_PRI_IN_RAW default-PRI finding
- static analyzer container: clean (scan-build: no bugs found)
- Ubuntu 26.04 devcontainer run-ci.sh: PASS, TOTAL 1283 / PASS 1190 / SKIP 93 / FAIL 0 / ERROR 0
